File: SSUMobile/Modules/Calendar/Model/SSUCalendarBuilder.m
//
//  SSUCalendarBuilder.m
//  SSUMobile
//
//  Created by Andrew Huss on 1/31/13.
//  Copyright (c) 2013 Sonoma State University Department of Computer Science. All rights reserved.
//

#import "SSUCalendarBuilder.h"
#import "SSUCalendarConstants.h"
#import "SSULogging.h"


NSString * const SSUCalendarEventKeyID = @"id";
NSString * const SSUCalendarEventKeyStart = @"start_date";
NSString * const SSUCalendarEventKeyEnd = @"end_date";
NSString * const SSUCalendarEventKeyTitle = @"title";
NSString * const SSUCalendarEventKeyOrganization = @"organization";
NSString * const SSUCalendarEventKeyCategory = @"category";
NSString * const SSUCalendarEventKeyLocation = @"location";
NSString * const SSUCalendarEventKeyDescription = @"description";
NSString * const SSUCalendarEventKeyImgURL = @"image_url";

@implementation SSUCalendarBuilder

+ (SSUEvent *) eventWithID:(NSNumber *)eventID inContext:(NSManagedObjectContext *)context {
    if (eventID == nil || eventID.integerValue == 0)
        return nil;
    
    BOOL created = NO;
    SSUEvent* event = (SSUEvent*)[self objectWithEntityName:SSUCalendarEntityEvent
                                                         ID:eventID
                                                    context:context
                                           entityWasCreated:&created];
    if (created) {
        event.id = eventID;
    }

    return event;
}

- (void) buildEvents:(NSArray *)events {
    SSULogDebug(@"Started events: %lu",(unsigned long)events.count);
    NSDate* buildStart = [NSDate date];
    
    for (NSDictionary * raw in events) {
        NSDictionary * eventData = [self cleanJSON:raw];
        NSNumber * eventId = @([eventData[SSUCalendarEventKeyID] integerValue]);
        SSUEvent * event = [SSUCalendarBuilder eventWithID:eventId inContext:self.context];
        
        event.title = eventData[SSUCalendarEventKeyTitle];
        event.startDate = [self.dateFormatter dateFromString:eventData[SSUCalendarEventKeyStart]];
        event.endDate = [self.dateFormatter dateFromString:eventData[SSUCalendarEventKeyEnd]];
        event.summary = eventData[SSUCalendarEventKeyDescription];
        event.location = eventData[SSUCalendarEventKeyLocation];
        event.imgURL = eventData[SSUCalendarEventKeyImgURL];
        event.organization = eventData[SSUCalendarEventKeyOrganization];
        event.category = eventData[SSUCalendarEventKeyCategory];
        
    }
    [self saveContext];
    
    SSULogDebug(@"Finished Events: %f",[[NSDate date] timeIntervalSinceDate:buildStart]);
}

@end

File: SSUMobile/Modules/Calendar/Model/SSUCalendarBuilder.swift
//
//  SSUCalendarBuilder.swift
//  SSUMobile
//
//  Created by Eric Amorde on 4/1/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation
import SwiftyJSON

class SSUCalendarBuilder: SSUMoonlightBuilder {

    private struct Keys {
        static let id = "id"
        static let startDate = "start_date"
        static let endDate = "end_date"
        static let title = "title"
        static let organization = "organization"
        static let category = "category"
        static let location = "location"
        static let summary = "description"
        static let imgURL = "image_url"
    }
    
    static func event(withID id: Int, inContext: NSManagedObjectContext) -> SSUEvent? {
        if id <= 0 {
            SSULogging.logError("Received invalid event id \(id)")
            return nil
        }
        
        let obj = self.object(withEntityName: "SSUEvent", id: id, context: inContext) as? SSUEvent
        // Here we don't know if this is a new object or one that already existed, so make sure it has an id
        obj?.id = Int32(id)
        
        return obj
    }
    
    override func build(_ results: Any!) {
        SSULogging.logDebug("Building events")
        let json = JSON(results)
        
        for entry in json.arrayValue {
            let mode = self.mode(fromJSONData: entry.dictionaryObject ?? [:])
            guard let event = SSUCalendarBuilder.event(withID: entry[Keys.id].intValue, inContext: self.context) else {
                SSULogging.logError("Unable to retrieve or create Event with id: \(entry[Keys.id].intValue)")
                return
            }
            if mode == .deleted {
                context.delete(event)
                continue
            }
            
            event.startDate = dateFormatter.date(from: entry[Keys.startDate].stringValue)
            event.endDate = dateFormatter.date(from: entry[Keys.endDate].stringValue)
            event.title = entry[Keys.title].string
            event.organization = entry[Keys.organization].string
            event.category = entry[Keys.category].string
            event.location = entry[Keys.location].string
            event.summary = entry[Keys.summary].string
        }
        saveContext()
        SSULogging.logDebug("Finish building events")
    }
}
